lua/smear_cursor/draw.lua

```diff
@@ -14,18 +14,67 @@ local MATRIX_CHARACTERS = { "▘", "▝", "▀", "▖", "▌", "▞", "▛", "
 -- Create a namespace for the extmarks
 local cursor_namespace = vim.api.nvim_create_namespace("smear_cursor")
 
-M.initialize_lists = function()
-	if _G.smear_cursor.window_ids == nil then
-		_G.smear_cursor.window_ids = { {} }
-		_G.smear_cursor.n_active_windows = { 0 }
+---@type table<number, {win:number, buf:number, used:boolean}[]>
+M.wins = {}
+
+-- remove any invalid windows
+function M.check_wins()
+	for tab in pairs(M.wins) do
+		M.wins[tab] = vim.tbl_filter(function(w)
+			return w.win and vim.api.nvim_win_is_valid(w.win) and w.buf and vim.api.nvim_buf_is_valid(w.buf)
+		end, M.wins[tab])
 	end
+end
+
+function M.get_win(tab, row, col)
+	M.wins[tab] = M.wins[tab] or {}
 
-	for _, tab in pairs(vim.api.nvim_list_tabpages()) do
-		if _G.smear_cursor.window_ids[tab] == nil then
-			_G.smear_cursor.window_ids[tab] = {}
-			_G.smear_cursor.n_active_windows[tab] = 0
+	for _, w in ipairs(M.wins[tab]) do
+		if not w.used then
```

Comment:
I had a `n_active_windows` counter (per tab, reset to 0 when `clean`-ing) to not have to loop through the whole list of active windows each time a character is added. What do you think?

Reply:
updated